### PR TITLE
Stand funnel tests

### PR DESCRIPTION
**Pull request by dmdashenkov**
Added `Given` util class for stand tests.
Added stand funnel tests covering:
 - Creation with various arguments.
 - Updates delivery from projection/aggregate repository into a stand.
 - Concurrent processing behavior.

### CODE REVIEW

**Comment by dmdashenkov** [comment]
@armiol PTAL

File: server/src/test/java/org/spine3/server/stand/Given.java [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+                                                             .setProjectId(ProjectId.newBuilder().setId("12345AD0"))
+                                                             .build())
+                                         .toBuilder()
+                                         .setTypeUrl(TypeUrl.SPINE_TYPE_URL_PREFIX + "/" + ProjectCreated.getDescriptor().getFullName())
+                                         .build())
+                    .setContext(EventContext.newBuilder()
+                                            .setDoNotEnrich(true)
+                                            .setCommandContext(CommandContext.getDefaultInstance())
+                                            .setEventId(EventId.newBuilder()
+                                                               .setUuid(PROJECT_UUID)
```
Comment (armiol) [inline server/src/test/java/org/spine3/server/stand/Given.java]: Why an event ID is built around project UUID?

File: server/src/test/java/org/spine3/server/stand/Given.java [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+        return new StandTestProjectionRepository(context);
+    }
+
+    /*package*/ static AggregateRepository<?, ?> aggregateRepo(BoundedContext context) {
+        return new StandTestAggregateRepository(context);
+    }
+
+    /*package*/ static BoundedContext boundedContext(Stand stand, int concurrentThreads) {
+        final Executor executor = concurrentThreads > 0 ? Executors.newFixedThreadPool(concurrentThreads) :
+                                  new Executor() { // Straightforward executor
```
Comment (armiol) [inline server/src/test/java/org/spine3/server/stand/Given.java]: Use MoreExecutors#directExecutor() instead.

File: server/src/test/java/org/spine3/server/stand/StandFunnelShould.java [hunk 3]
```diff
@@ -53,6 +70,56 @@ public void initialize_properly_with_stand_only() {
         Assert.assertNotNull(standFunnel);
     }
 
+    @Test
+    public void initialize_properly_with_various_builder_options() {
```
Comment (armiol) [inline server/src/test/java/org/spine3/server/stand/StandFunnelShould.java]: Please split into a number of tests.

File: server/src/test/java/org/spine3/server/stand/StandFunnelShould.java [hunk 4]
```diff
@@ -53,6 +70,56 @@ public void initialize_properly_with_stand_only() {
...
+
+
+        final StandFunnel emptyExecutorFunnel = StandFunnel.newBuilder()
+                                                           .setStand(TestStandFactory.create())
+                                                           .setExecutor(new Executor() {
+                                                               @Override
+                                                               public void execute(Runnable neverCalled) { }
+                                                           })
+                                                           .build();
+        Assert.assertNotNull(emptyExecutorFunnel);
```
Comment (armiol) [inline server/src/test/java/org/spine3/server/stand/StandFunnelShould.java]: You don't test a plain StandFunnel.Builder with no executor set.

File: server/src/test/java/org/spine3/server/stand/StandFunnelShould.java [hunk 5]
```diff
@@ -53,6 +70,56 @@ public void initialize_properly_with_stand_only() {
...
+        final StandFunnel funnelForBusyStand = StandFunnel.newBuilder()
+                                                          .setStand(stand)
+                                                          .setExecutor(Executors.newSingleThreadExecutor())
+                                                          .build();
+        Assert.assertNotNull(funnelForBusyStand);
+
+
+        final StandFunnel emptyExecutorFunnel = StandFunnel.newBuilder()
+                                                           .setStand(TestStandFactory.create())
+                                                           .setExecutor(new Executor() {
```
Comment (armiol) [inline server/src/test/java/org/spine3/server/stand/StandFunnelShould.java]: Also, I don't see a difference between initialisation with different executors. These cases will differ in an operation behaviour, not in the initialisation.

File: server/src/test/java/org/spine3/server/stand/StandFunnelShould.java [hunk 6]
```diff
@@ -80,9 +147,14 @@ public void execute(Runnable command) {
 
     // **** Negative scenarios (unit) ****
 
-    /**
-     * - Fail to initialise with improper stand.
-     */
+    @SuppressWarnings("ResultOfMethodCallIgnored")
+    @Test(expected = NullPointerException.class)
+    public void fail_to_initialize_with_inproper_stand() {
```
Comment (armiol) [inline server/src/test/java/org/spine3/server/stand/StandFunnelShould.java]: `improper`.

File: server/src/test/java/org/spine3/server/stand/StandFunnelShould.java [hunk 7]
```diff
@@ -80,9 +147,14 @@ public void execute(Runnable command) {
...
+        @SuppressWarnings("ConstantConditions")
```
Comment (armiol) [inline server/src/test/java/org/spine3/server/stand/StandFunnelShould.java]: Please check if this `@SuppressWarnings` statement makes sense.

File: server/src/test/java/org/spine3/server/stand/StandFunnelShould.java [hunk 8]
```diff
@@ -94,10 +166,150 @@ public void fail_to_initialize_from_empty_builder() {
     // **** Integration scenarios (<source> -> StandFunnel -> Mock Stand) ****
 
     /**
-     * - Deliver updates from projection repo on update;
-     * - deliver updates from aggregate repo on update;
+     * - Deliver updates from projection projectionRepo on update;
```
Comment (armiol) [inline server/src/test/java/org/spine3/server/stand/StandFunnelShould.java]: Let's remove this section at all. You substitute it with actual tests.

File: server/src/test/java/org/spine3/server/stand/StandFunnelShould.java [hunk 9]
```diff
@@ -94,10 +166,150 @@ public void fail_to_initialize_from_empty_builder() {
...
+        final Random random = new SecureRandom();
+
+        for (int i = 0; i < result.length; i++) {
+            result[i] = random.nextBoolean() ? aggregateRepositoryDispatch() : projectionRepositoryDispatch();
+        }
+
+        return result;
+    }
+
+    private static void deliverUpdates(boolean isMultiThread, BoundedContextAction... dispatchActions) {
```
Comment (armiol) [inline server/src/test/java/org/spine3/server/stand/StandFunnelShould.java]: `checkUpdateDelivery()`.

File: server/src/test/java/org/spine3/server/stand/StandFunnelShould.java [hunk 10]
```diff
@@ -94,10 +166,150 @@ public void fail_to_initialize_from_empty_builder() {
...
+        final Stand stand = mock(Stand.class);
+        final BoundedContext boundedContext = spy(Given.boundedContext(stand,
+                                                                       isMultiThread ?
+                                                                       Given.THREADS_COUNT_IN_POOL_EXECUTOR : 0));
+
+        for (BoundedContextAction dispatchAction : dispatchActions) {
+            dispatchAction.perform(boundedContext);
+        }
+
+        // Was called as much times as there are dispatch actions.
```
Comment (armiol) [inline server/src/test/java/org/spine3/server/stand/StandFunnelShould.java]: `as many times`

File: server/src/test/java/org/spine3/server/stand/StandFunnelShould.java (hunk 9 above)
Comment (armiol) [inline server/src/test/java/org/spine3/server/stand/StandFunnelShould.java]: `isMultiThreaded` or `isConcurrent`.

File: server/src/test/java/org/spine3/server/stand/StandFunnelShould.java [hunk 11]
```diff
@@ -94,10 +166,150 @@ public void fail_to_initialize_from_empty_builder() {
...
+            public void perform(BoundedContext context) {
+                // Init repository
+                final AggregateRepository<?, ?> repository = Given.aggregateRepo(context);
+
+                repository.initStorage(InMemoryStorageFactory.getInstance());
+
+                try {
+                    repository.dispatch(Given.validCommand());
+                } catch (IllegalStateException e) {
+                    // Handle null event dispatch after command handling.
```
Comment (armiol) [inline server/src/test/java/org/spine3/server/stand/StandFunnelShould.java]: `..dispatching after the command is handled.`

File: server/src/test/java/org/spine3/server/stand/StandFunnelShould.java [hunk 12]
```diff
@@ -94,10 +166,150 @@ public void fail_to_initialize_from_empty_builder() {
...
+                    Assert.assertTrue(e.getMessage().contains("No record found for command ID: EMPTY"));
```
Comment (armiol) [inline server/src/test/java/org/spine3/server/stand/StandFunnelShould.java]: Why do we check that? That is not related to StandFunnel functionality.

Also, it does not read: you dispatch a `Given.validCommand()`; in this case you should have returned some event, not raise an exception.

File: server/src/test/java/org/spine3/server/stand/StandFunnelShould.java [hunk 13]
```diff
@@ -94,10 +166,150 @@ public void fail_to_initialize_from_empty_builder() {
...
+
+
+    @SuppressWarnings("MethodWithMultipleLoops")
+    @Test
+    public void deliver_updates_through_several_threads() throws InterruptedException {
+        final int threadsCount = Given.THREADS_COUNT_IN_POOL_EXECUTOR;
+        @SuppressWarnings("LocalVariableNamingConvention") // Too long variable name
+        final int threadExecutionMaxAwaitSeconds = Given.AWAIT_SECONDS;
+
+        final Map<String, Object> threadInvocationRegistry = new ConcurrentHashMap<>(threadsCount);
```
Comment (armiol) [inline server/src/test/java/org/spine3/server/stand/StandFunnelShould.java]: Why do you need `Map`? Seems that `Set` is enough.

File: server/src/test/java/org/spine3/server/stand/StandFunnelShould.java [hunk 14]
```diff
@@ -94,10 +166,150 @@ public void fail_to_initialize_from_empty_builder() {
...
+
+        final Stand stand = mock(Stand.class);
+        doNothing().when(stand).update(ArgumentMatchers.any(), any(Any.class));
+
+        final StandFunnel standFunnel = StandFunnel.newBuilder()
+                                                   .setStand(stand)
+                                                   .build();
+
+        final ExecutorService processes = Executors.newFixedThreadPool(threadsCount);
```
Comment (armiol) [inline server/src/test/java/org/spine3/server/stand/StandFunnelShould.java]: This is not `processes`. This is an `executor`.

File: server/src/main/java/org/spine3/server/stand/StandFunnel.java [hunk 15]
```diff
@@ -109,6 +109,7 @@ public Stand getStand() {
          * <p> The value must not be null.
          *
          * <p> If this method is not used, a default value will be used.
+         * // TODO:13-09-16:dmytro.dashenkov: Correct docs. No default value for Stand is used, null value leads to a fail instead.
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/stand/StandFunnel.java]: Please correct the docs yourselves. That's a nice catch. 
Of course, `Stand` instance must be set.

Comment (armiol) [inline server/src/main/java/org/spine3/server/stand/StandFunnel.java]: Let's discuss both of your `TODO`s.

**Comment by armiol** [comment]
@dmdashenkov please see my comments.

**Comment by armiol** [comment]
@dmdashenkov LGTM. Please also check the coverage of `StandFunnel` [here](https://codecov.io/gh/SpineEventEngine/core-java/branch/stand-funnel-tests). We need to be as close to 100% coverage as possible.

**Comment by dmdashenkov** [comment]
@armiol For now both Covecov and IDEA's "run with coverage" reports 100% coverage of `StandFunnel`. Note, that this class has only one public method.
